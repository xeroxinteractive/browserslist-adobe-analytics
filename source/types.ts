import { JWTAuthConfig } from '@adobe/jwt-auth';

export interface BaseOptions extends Omit<JWTAuthConfig, 'metaScopes'> {
  rsid: string;
  globalId: string;
}

export interface WriteOptions extends BaseOptions {
  cwd?: string;
  filename?: string;
}

export interface RankedReportData {
  totalPages: number;
  firstPage: boolean;
  lastPage: boolean;
  numberOfElements: number;
  number: number;
  totalElements: number;
  message?: string;
  request?: object;
  columns: RankedColumnMetaData;
  rows: Row[];
  summaryData: RankedSummaryData;
}

export interface RankedColumnMetaData {
  dimension: ReportDimension;
  columnIds: string[];
  columnErrors?: RankedColumnError[];
}

export interface ReportDimension {
  id: string;
  type: ReportDimensionType;
}

export enum ReportDimensionType {
  STRING = 'string',
}

export interface RankedColumnError {
  columnId: string;
  errorCode:
    | 'unauthorized_metric'
    | 'unauthorized_dimension'
    | 'unauthorized_dimension_global'
    | 'anomaly_detection_failure_unexpected_item_count'
    | 'anomaly_detection_failure_tsa_service'
    | 'not_enabled_metric'
    | 'not_enabled_dimension'
    | 'not_enabled_dimension_global';
  errorId: string;
  errorDescription: string;
}

export interface Row {
  itemId: string;
  value?: string;
  data: number[];
}

export interface RankedSummaryData {
  filteredTotals: number[];
  totals: number[];
  'col-max': number[];
  'col-min': number[];
}

/**
 * Thrown if node-fetch response status is not >= 200 < 300.
 */
export class ResponseError extends Error {
  public status: number;
  /**
   * Creates a new ResponseError instance.
   *
   * @param statusText - The message returned by the server.
   * @param status - The HTTP status code returned by the server.
   */
  public constructor(statusText: string, status: number) {
    super(`${statusText} (${status})`);
    this.status = status;
    this.name = this.constructor.name;
    Object.setPrototypeOf(this, new.target.prototype);
    Error.captureStackTrace(this, this.constructor);
  }
}
