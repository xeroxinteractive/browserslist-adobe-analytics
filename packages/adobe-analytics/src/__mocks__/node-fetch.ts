const { FetchError } = jest.requireActual('node-fetch');

import mockBrowserReport from '../__specs__/browser-report';

const fetch = jest.fn(
  async (): Promise<Record<string, any>> => ({
    json: async (): Promise<Record<string, any>> => {
      return mockBrowserReport;
    },
    ok: true,
  })
);

export { FetchError };

export default fetch;
