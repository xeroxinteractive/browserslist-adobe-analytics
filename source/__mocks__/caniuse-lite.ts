export const agents = {
  ie: {
    release_date: {
      '6': 998870400,
      '7': 1161129600,
      '8': 1237420800,
      '9': 1300060800,
      '10': 1346716800,
      '11': 1381968000,
      '5.5': 962323200,
    },
  },
  edge: {
    release_date: {
      '12': 1438128000,
      '13': 1447286400,
      '14': 1470096000,
      '15': 1491868800,
      '16': 1508198400,
      '17': 1525046400,
      '18': 1542067200,
      '79': 1579046400,
      '80': 1581033600,
      '81': 1586736000,
    },
  },
  firefox: {
    release_date: {
      '2': 1161648000,
      '3': 1213660800,
      '4': 1300752000,
      '5': 1308614400,
      '6': 1313452800,
      '7': 1317081600,
      '8': 1317081600,
      '9': 1320710400,
      '10': 1324339200,
      '11': 1327968000,
      '12': 1331596800,
      '13': 1335225600,
      '14': 1338854400,
      '15': 1342483200,
      '16': 1346112000,
      '17': 1349740800,
      '18': 1353628800,
      '19': 1357603200,
      '20': 1361232000,
      '21': 1364860800,
      '22': 1368489600,
      '23': 1372118400,
      '24': 1375747200,
      '25': 1379376000,
      '26': 1386633600,
      '27': 1391472000,
      '28': 1395100800,
      '29': 1398729600,
      '30': 1402358400,
      '31': 1405987200,
      '32': 1409616000,
      '33': 1413244800,
      '34': 1417392000,
      '35': 1421107200,
      '36': 1424736000,
      '37': 1428278400,
      '38': 1431475200,
      '39': 1435881600,
      '40': 1439251200,
      '41': 1442880000,
      '42': 1446508800,
      '43': 1450137600,
      '44': 1453852800,
      '45': 1457395200,
      '46': 1461628800,
      '47': 1465257600,
      '48': 1470096000,
      '49': 1474329600,
      '50': 1479168000,
      '51': 1485216000,
      '52': 1488844800,
      '53': 1492560000,
      '54': 1497312000,
      '55': 1502150400,
      '56': 1506556800,
      '57': 1510617600,
      '58': 1516665600,
      '59': 1520985600,
      '60': 1525824000,
      '61': 1529971200,
      '62': 1536105600,
      '63': 1540252800,
      '64': 1544486400,
      '65': 1548720000,
      '66': 1552953600,
      '67': 1558396800,
      '68': 1562630400,
      '69': 1567468800,
      '70': 1571788800,
      '71': 1575331200,
      '72': 1578355200,
      '73': 1581379200,
      '74': 1583798400,
      '75': 1586304000,
      '76': 1588636800,
      '77': null,
      '78': null,
      '3.5': 1246320000,
      '3.6': 1264032000,
    },
  },
  chrome: {
    release_date: {
      '4': 1264377600,
      '5': 1274745600,
      '6': 1283385600,
      '7': 1287619200,
      '8': 1291248000,
      '9': 1296777600,
      '10': 1299542400,
      '11': 1303862400,
      '12': 1307404800,
      '13': 1312243200,
      '14': 1316131200,
      '15': 1316131200,
      '16': 1319500800,
      '17': 1323734400,
      '18': 1328659200,
      '19': 1332892800,
      '20': 1337040000,
      '21': 1340668800,
      '22': 1343692800,
      '23': 1348531200,
      '24': 1352246400,
      '25': 1357862400,
      '26': 1361404800,
      '27': 1364428800,
      '28': 1369094400,
      '29': 1374105600,
      '30': 1376956800,
      '31': 1384214400,
      '32': 1389657600,
      '33': 1392940800,
      '34': 1397001600,
      '35': 1400544000,
      '36': 1405468800,
      '37': 1409011200,
      '38': 1412640000,
      '39': 1416268800,
      '40': 1421798400,
      '41': 1425513600,
      '42': 1429401600,
      '43': 1432080000,
      '44': 1437523200,
      '45': 1441152000,
      '46': 1444780800,
      '47': 1449014400,
      '48': 1453248000,
      '49': 1456963200,
      '50': 1460592000,
      '51': 1464134400,
      '52': 1469059200,
      '53': 1472601600,
      '54': 1476230400,
      '55': 1480550400,
      '56': 1485302400,
      '57': 1489017600,
      '58': 1492560000,
      '59': 1496707200,
      '60': 1500940800,
      '61': 1504569600,
      '62': 1508198400,
      '63': 1512518400,
      '64': 1516752000,
      '65': 1520294400,
      '66': 1523923200,
      '67': 1527552000,
      '68': 1532390400,
      '69': 1536019200,
      '70': 1539648000,
      '71': 1543968000,
      '72': 1548720000,
      '73': 1552348800,
      '74': 1555977600,
      '75': 1559606400,
      '76': 1564444800,
      '77': 1568073600,
      '78': 1571702400,
      '79': 1575936000,
      '80': 1580860800,
      '81': 1586304000,
      '83': null,
      '84': null,
      '85': null,
    },
  },
  safari: {
    release_date: {
      '4': 1244419200,
      '5': 1275868800,
      '6': 1343174400,
      '7': 1382400000,
      '8': 1413417600,
      '9': 1443657600,
      '10': 1474329600,
      '11': 1505779200,
      '12': 1537142400,
      '13': 1568851200,
      '3.1': 1205798400,
      '3.2': 1226534400,
      '5.1': 1311120000,
      '6.1': 1382400000,
      '7.1': 1410998400,
      '9.1': 1458518400,
      '10.1': 1490572800,
      '11.1': 1522281600,
      '12.1': 1553472000,
      '13.1': 1585008000,
      TP: null,
    },
  },
  opera: {
    release_date: {
      '9': 1150761600,
      '11': 1292457600,
      '12': 1323129600,
      '15': 1372723200,
      '16': 1377561600,
      '17': 1381104000,
      '18': 1386288000,
      '19': 1390867200,
      '20': 1393891200,
      '21': 1399334400,
      '22': 1401753600,
      '23': 1405987200,
      '24': 1409616000,
      '25': 1413331200,
      '26': 1417132800,
      '27': 1422316800,
      '28': 1425945600,
      '29': 1430179200,
      '30': 1433808000,
      '31': 1438646400,
      '32': 1442448000,
      '33': 1445904000,
      '34': 1449100800,
      '35': 1454371200,
      '36': 1457308800,
      '37': 1462320000,
      '38': 1465344000,
      '39': 1470096000,
      '40': 1474329600,
      '41': 1477267200,
      '42': 1481587200,
      '43': 1486425600,
      '44': 1490054400,
      '45': 1494374400,
      '46': 1498003200,
      '47': 1502236800,
      '48': 1506470400,
      '49': 1510099200,
      '50': 1515024000,
      '51': 1517961600,
      '52': 1521676800,
      '53': 1525910400,
      '54': 1530144000,
      '55': 1534982400,
      '56': 1537833600,
      '57': 1543363200,
      '58': 1548201600,
      '60': 1554768000,
      '62': 1561593600,
      '63': 1566259200,
      '64': 1570406400,
      '65': 1573689600,
      '66': 1578441600,
      '67': 1583971200,
      '68': 1587513600,
      '9.5-9.6': 1223424000,
      '10.0-10.1': 1251763200,
      '10.5': 1267488000,
      '10.6': 1277942400,
      '11.1': 1302566400,
      '11.5': 1309219200,
      '11.6': 1323129600,
      '12.1': 1352073600,
    },
  },
  ios_saf: {
    release_date: {
      '8': 1410912000,
      '3.2': 1270252800,
      '4.0-4.1': 1283904000,
      '4.2-4.3': 1299628800,
      '5.0-5.1': 1331078400,
      '6.0-6.1': 1359331200,
      '7.0-7.1': 1394409600,
      '8.1-8.4': 1413763200,
      '9.0-9.2': 1442361600,
      '9.3': 1458518400,
      '10.0-10.2': 1473724800,
      '10.3': 1490572800,
      '11.0-11.2': 1505779200,
      '11.3-11.4': 1522281600,
      '12.0-12.1': 1537142400,
      '12.2-12.4': 1553472000,
      '13.0-13.1': 1568851200,
      '13.2': 1572220800,
      '13.3': 1580169600,
      '13.4': 1585008000,
    },
  },
  op_mini: {
    release_date: { all: 1426464000 },
  },
  android: {
    release_date: {
      '3': 1298332800,
      '4': 1318896000,
      '81': 1587427200,
      '2.1': 1256515200,
      '2.2': 1274313600,
      '2.3': 1291593600,
      '4.1': 1341792000,
      '4.2-4.3': 1374624000,
      '4.4': 1386547200,
      '4.4.3-4.4.4': 1401667200,
    },
  },
  bb: {
    release_date: { '7': 1325376000, '10': 1359504000 },
  },
  op_mob: {
    release_date: {
      '10': 1287100800,
      '11': 1300752000,
      '12': 1330300800,
      '46': 1474588800,
      '11.1': 1314835200,
      '11.5': 1318291200,
      '12.1': 1349740800,
    },
  },
  and_chr: {
    release_date: { '81': 1587427200 },
  },
  and_ff: {
    release_date: { '68': 1567468800 },
  },
  ie_mob: {
    release_date: { '10': 1340150400, '11': 1353456000 },
  },
  and_uc: {
    release_date: { '12.12': 1471392000 },
  },
  samsung: {
    release_date: {
      '4': 1461024000,
      '5.0-5.4': 1481846400,
      '6.2-6.4': 1509408000,
      '7.2-7.4': 1528329600,
      '8.2': 1546128000,
      '9.2': 1554163200,
      '10.1': 1567900800,
      '11.1': 1582588800,
    },
  },
  and_qq: {
    release_date: { '1.2': 1483228800 },
  },
  baidu: {
    release_date: { '7.12': 1491004800 },
  },
  kaios: {
    release_date: { '2.5': 1527811200 },
  },
};
