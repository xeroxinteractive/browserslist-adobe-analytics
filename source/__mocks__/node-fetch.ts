const { FetchError } = require.requireActual('node-fetch');

import mockBrowserReport from '../__specs__/browser-report';

const fetch = jest.fn(
  async (): Promise<object> => ({
    json: async (): Promise<object> => {
      return mockBrowserReport;
    },
    ok: true,
  })
);

export { FetchError };

export default fetch;
