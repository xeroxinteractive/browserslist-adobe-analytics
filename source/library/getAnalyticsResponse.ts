import authorize, { JWTAuthConfig } from '@adobe/jwt-auth';
import { BaseOptions, ResponseError, RankedReportData } from '../types';
import fetch from 'node-fetch';
import getRequestBody from './getRequestBody';

/**
 * Pulls browser data from Adobe Analytics.
 *
 * @param options - Adobe Analytics credential options.
 * @returns Browser data.
 */
export default async function getAnalyticsResponse(
  options: BaseOptions
): Promise<RankedReportData | undefined> {
  const config: JWTAuthConfig = {
    ...options,
    metaScopes: ['ent_analytics_bulk_ingest_sdk'],
  };
  try {
    const { access_token } = await authorize(config);
    const response = await fetch(
      `https://analytics.adobe.io/api/${options.globalId}/reports`,
      {
        method: 'post',
        body: JSON.stringify(getRequestBody(options)),
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Bearer ${access_token}`,
          'x-proxy-global-company-id': options.globalId,
          'x-api-key': options.clientId,
        },
      }
    );
    if (!response.ok) {
      throw new ResponseError(response.statusText, response.status);
    }
    return await response.json();
  } catch (e) {
    console.error(e);
  }
}
