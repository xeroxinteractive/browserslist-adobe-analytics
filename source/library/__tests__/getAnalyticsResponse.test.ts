jest.mock('util');

import { mocked } from 'ts-jest/utils';
import nodeFetch, { Response } from 'node-fetch';
import authenticate from '@adobe/jwt-auth';
import mockBrowserReport from '../../__specs__/browser-report';
import mockOptions from '../../__specs__/options';
import * as util from 'util';
import * as MockDate from 'mockdate';
// eslint-disable-next-line jest/no-mocks-import
import { FetchError } from '../../__mocks__/node-fetch';
import { ResponseError } from '../../types';

const mockFetch = mocked(nodeFetch, true);
const mockAuthenticate = mocked(authenticate, true);
MockDate.set('2019-11-01T00:00:00.000');
const mockReadFile = jest.fn();
mocked(util).promisify.mockImplementation(() => mockReadFile);

import getAnalyticsResponse from '../getAnalyticsResponse';

beforeEach(() => {
  jest.clearAllMocks();
});

test('normal behaviour', async () => {
  await expect(getAnalyticsResponse(mockOptions)).resolves.toEqual(
    mockBrowserReport
  );
  expect(mockFetch).toHaveBeenCalledTimes(1);
  expect(mockFetch.mock.calls[0]).toMatchSnapshot();
  expect(mockAuthenticate).toHaveBeenCalledTimes(1);
  expect(mockAuthenticate).toHaveBeenCalledWith({
    ...mockOptions,
    metaScopes: ['ent_analytics_bulk_ingest_sdk'],
  });
  expect(mockReadFile).not.toHaveBeenCalled();
});

test('HTTP error', async () => {
  mockFetch.mockImplementationOnce(
    async (): Promise<Response> =>
      (({
        json: async (): Promise<Response> => {
          return undefined as any;
        },
        ok: false,
        statusText: 'Forbidden',
        status: 403,
      } as Partial<Response>) as Response)
  );
  await expect(getAnalyticsResponse(mockOptions)).rejects.toThrow(
    new ResponseError('Forbidden', 403)
  );
  expect(mockFetch).toHaveBeenCalledTimes(1);
  expect(mockAuthenticate).toHaveBeenCalledTimes(1);
  expect(mockAuthenticate).toHaveBeenCalledWith({
    ...mockOptions,
    metaScopes: ['ent_analytics_bulk_ingest_sdk'],
  });
  expect(mockReadFile).not.toHaveBeenCalled();
});

test('Fetch error', async () => {
  const mockFetchError = new FetchError('---message---', '---type---');
  mockFetch.mockImplementationOnce(
    async (): Promise<Response> =>
      (({
        json: async (): Promise<Response> => {
          throw mockFetchError;
        },
        ok: true,
      } as Partial<Response>) as Response)
  );
  await expect(getAnalyticsResponse(mockOptions)).rejects.toThrow(
    mockFetchError
  );
  expect(mockFetch).toHaveBeenCalledTimes(1);
  expect(mockAuthenticate).toHaveBeenCalledTimes(1);
  expect(mockAuthenticate).toHaveBeenCalledWith({
    ...mockOptions,
    metaScopes: ['ent_analytics_bulk_ingest_sdk'],
  });
  expect(mockReadFile).not.toHaveBeenCalled();
});

test('private key from file', async () => {
  mockReadFile.mockImplementationOnce(() => {
    return '---private-key-from-file---';
  });
  await expect(
    getAnalyticsResponse({
      ...mockOptions,
      privateKey: undefined,
      privateKeyPath: '/path/to/private.key',
    })
  ).resolves.toEqual(mockBrowserReport);
  expect(mockFetch).toHaveBeenCalledTimes(1);
  expect(mockFetch.mock.calls[0]).toMatchSnapshot();
  expect(mockAuthenticate).toHaveBeenCalledTimes(1);
  expect(mockAuthenticate).toHaveBeenCalledWith({
    ...mockOptions,
    privateKey: '---private-key-from-file---',
    privateKeyPath: '/path/to/private.key',
    metaScopes: ['ent_analytics_bulk_ingest_sdk'],
  });
  expect(mockReadFile).toHaveBeenCalledTimes(1);
});

test('no private key', async () => {
  await expect(
    getAnalyticsResponse({
      ...mockOptions,
      privateKey: undefined,
      privateKeyPath: '/path/to/private.key',
    })
  ).rejects.toThrow(
    new Error(
      'Invalid private key either pass the raw key via `privateKey` or a path to it via `privateKeyPath`.'
    )
  );
  expect(mockFetch).not.toHaveBeenCalled();
  expect(mockAuthenticate).not.toHaveBeenCalled();
  expect(mockReadFile).toHaveBeenCalledTimes(1);
});

test('no private or private key path', async () => {
  await expect(
    getAnalyticsResponse({
      ...mockOptions,
      privateKey: undefined,
    } as any)
  ).rejects.toThrow(
    new Error(
      'Invalid private key either pass the raw key via `privateKey` or a path to it via `privateKeyPath`.'
    )
  );
  expect(mockFetch).not.toHaveBeenCalled();
  expect(mockAuthenticate).not.toHaveBeenCalled();
  expect(mockReadFile).not.toHaveBeenCalled();
});
