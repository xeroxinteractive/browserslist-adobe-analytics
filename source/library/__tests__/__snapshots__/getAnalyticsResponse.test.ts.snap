// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normal behaviour 1`] = `
Array [
  "https://analytics.adobe.io/api/---global-id---/reports",
  Object {
    "body": "{\\"rsid\\":\\"---rsid---\\",\\"globalFilters\\":[{\\"type\\":\\"dateRange\\",\\"dateRange\\":\\"2019-08-01T00:00:00.000/2019-11-01T00:00:00.000\\"}],\\"metricContainer\\":{\\"metrics\\":[{\\"columnId\\":\\"0\\",\\"id\\":\\"metrics/visitors\\"}]},\\"dimension\\":\\"variables/browser\\",\\"settings\\":{\\"countRepeatInstances\\":true,\\"limit\\":50,\\"page\\":0,\\"nonesBehavior\\":\\"return-nones\\"}}",
    "headers": Object {
      "Authorization": "Bearer ---access-token---",
      "Content-Type": "application/json",
      "x-api-key": "---client-id---",
      "x-proxy-global-company-id": "---global-id---",
    },
    "method": "post",
  },
]
`;
